File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:

            m = l + (r-l) // 2
            if nums[m] == target:
                return m
            
            # Is left half sorted?
            if nums[l] <= nums[m]:
                if nums[l] <= target <= nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                # right half sorted
                if nums[m] <= target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1
        

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.cpp
class Solution {
public:
    int bSearch(vector<int>& nums, int l, int r, int target) {
        while (l <= r) {
            int m = l + (r-l)/2;
            if (nums[m] == target) {
                return m;
            }
            else if (nums[m] < target) {
                l = m + 1;
            } else {
                r = m - 1;
            }
        }
        return -1;
    }
    int search(vector<int>& nums, int target) {

        // find pivot
        int l = 0, r = nums.size()-1;
        while (l < r) {
            int m = l + (r-l)/2;
            if (nums[m] > nums[r]) {
                l = m + 1;
            } else {
                r = m;
            }
        }

        int pivot = l;
        // serach left or right of pivot
        if (nums[pivot] <= target && target <= nums.back()) {
            // Serach right
            l = pivot;
            r = nums.size()-1;
            
        } else {
            // Search left
            l = 0;
            r = pivot-1;
        }

        return bSearch(nums, l, r, target);
    }
};